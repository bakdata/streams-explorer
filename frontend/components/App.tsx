import { DownOutlined, LoadingOutlined } from "@ant-design/icons";
import {
  Alert,
  Button,
  Dropdown,
  Layout,
  Menu,
  message,
  Row,
  Spin,
} from "antd";
import { useRouter } from "next/router";
import React, { useCallback, useEffect, useRef, useState } from "react";
import { useResizeDetector } from "react-resize-detector";
import { useMutate } from "restful-react";
import {
  HTTPValidationError,
  useGetMetricsApiMetricsGet,
  useGetPipelinesApiPipelinesGet,
  useGetPositionedGraphApiGraphGet,
} from "./api/fetchers";
import DetailsCard from "./DetailsCard";
import GraphVisualization from "./GraphVisualization";
import Node from "./Node";
import Search from "./Search";
import Settings from "./Settings";

const { Header, Content } = Layout;

const App: React.FC = () => {
  const ALL_PIPELINES = "all pipelines";
  const [currentPipeline, setCurrentPipeline] = useState(ALL_PIPELINES);
  const [detailNode, setDetailNode] = useState<Node>();
  const [focusedNode, setFocusedNode] = useState<Node>();
  const router = useRouter();
  const { query } = router;
  const ref = useRef<HTMLDivElement>(null);
  const onResize = useCallback(() => {}, []);
  const { width, height } = useResizeDetector({
    targetRef: ref,
    refreshMode: "debounce",
    refreshRate: 100,
    onResize,
  });
  const DEFAULT_REFRESH_INTERVAL = 30;
  const refreshIntervals: Record<number, string> = {
    0: "off",
    60: "60s",
    30: "30s",
    10: "10s",
  };
  const REFRESH_INTERVAL = "metrics-interval";
  const [refreshInterval, setRefreshInterval] = useState(0);
  const [animate, setAnimate] = useState(true);

  // on initial page load
  useEffect(() => {
    const storedRefreshInterval = Number(
      localStorage.getItem(REFRESH_INTERVAL) || DEFAULT_REFRESH_INTERVAL
    );
    setRefreshInterval(storedRefreshInterval);
    if (storedRefreshInterval) {
      refetchMetrics();
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    localStorage.setItem(REFRESH_INTERVAL, refreshInterval.toString());
  }, [refreshInterval]);

  const { mutate: update, loading: isUpdating } = useMutate({
    verb: "POST",
    path: "/api/update",
  });

  const {
    data: graph,
    loading: isLoadingGraph,
    error: graphError,
    refetch: graphRefetch,
  } = useGetPositionedGraphApiGraphGet({
    queryParams: currentPipeline !== ALL_PIPELINES
      ? { pipeline_name: currentPipeline }
      : undefined,
  });

  const {
    refetch: retryPipelineGraph,
    error: retryPipelineGraphError,
    data: retryPipelineGraphData,
  } = useGetPositionedGraphApiGraphGet({
    queryParams: { pipeline_name: currentPipeline },
    lazy: true,
  });

  const {
    data: pipelines,
    loading: isLoadingPipelines,
    error: pipelineError,
  } = useGetPipelinesApiPipelinesGet({});

  const {
    data: metrics,
    loading: isLoadingMetrics,
    refetch: refetchMetrics,
    error: metricsError,
  } = useGetMetricsApiMetricsGet({ lazy: true });

  useEffect(() => {
    if (refreshInterval && refreshInterval > 0) {
      const interval = setInterval(refetchMetrics, refreshInterval * 1000);
      return () => clearInterval(interval);
    }
  }, [refreshInterval]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (!graph) return;

    const pipeline = query.pipeline as string;
    if (pipeline) {
      setCurrentPipeline(pipeline);
    }

    const focusNode = query["focus-node"] as string;
    if (focusNode) {
      const node = graph?.nodes.find((node) => node.id === focusNode);
      if (node) {
        setDetailNode(node);
        setFocusedNode(node);
      }
    }
  }, [graph, query]);

  useEffect(() => {
    if (graphError) {
      let errorMessage: string | undefined;
      if ("data" in graphError) {
        // specific pipeline was not found
        const data = graphError["data"] as HTTPValidationError;
        if (data.detail) {
          errorMessage = data.detail.toString();
        }
      }
      message.error(errorMessage || "Failed loading graph", 5);

      if (graphError.status === 404 && currentPipeline !== ALL_PIPELINES) {
        // check if a re-scrape solves it
        const hideMessage = message.warning("Refreshing pipelines", 0);
        update({})
          .then(() => {
            retryPipelineGraph();
          })
          .catch(() => {
            redirectAllPipelines();
          })
          .finally(() => {
            hideMessage();
          });
      }
    }
  }, [graphError]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (
      retryPipelineGraphError
      && retryPipelineGraphError.status === 404
      && currentPipeline !== ALL_PIPELINES
    ) {
      // pipeline still not found
      redirectAllPipelines();
    } else if (retryPipelineGraphData) {
      message.success("Found pipeline!");
      graphRefetch();
    } // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [retryPipelineGraphError, retryPipelineGraphData]);

  const redirectAllPipelines = () => {
    message.info("Redirecting to all pipelines");
    setCurrentPipeline(ALL_PIPELINES);
    router.push("/");
  };

  useEffect(() => {
    if (metricsError) {
      message.warning("Failed fetching metrics");
    }
  }, [metricsError]);

  useEffect(() => {
    if (pipelineError) {
      message.error("Failed loading pipeline names");
    }
  }, [pipelineError]);

  const menuPipeline = (
    <Menu
      data-testid="pipeline-select"
      onClick={(e) => {
        let pipeline = e.key.toString();
        if (pipeline === ALL_PIPELINES) {
          router.push("/");
        } else {
          router.push(`/?pipeline=${e.key.toString()}`);
        }
        setCurrentPipeline(e.key.toString());
        setFocusedNode(undefined);
      }}
    >
      <Menu.Item key={ALL_PIPELINES}>
        <i>{ALL_PIPELINES}</i>
      </Menu.Item>
      {pipelines?.pipelines.map((name: string) => (
        <Menu.Item data-testid="pipeline-option" key={name}>
          {name}
        </Menu.Item>
      ))}
    </Menu>
  );

  const menuRefresh = (
    <Menu
      data-testid="metrics-select"
      onClick={(e) => {
        setRefreshInterval(Number(e.key));
      }}
    >
      {Object.keys(refreshIntervals).map((key: any) => (
        <Menu.Item key={key}>{refreshIntervals[key]}</Menu.Item>
      ))}
    </Menu>
  );

  if (
    typeof window !== undefined
    && !isLoadingGraph
    && !isLoadingPipelines
    && !isUpdating
  ) {
    return (
      <div ref={ref} className="application">
        <Layout className="layout">
          <Header className="header" style={{ zIndex: 2 }}>
            <Menu theme="dark" mode="horizontal" selectable={false}>
              <Menu.Item key="1">
                Pipeline:&nbsp;
                <Dropdown overlay={menuPipeline} placement="bottomLeft" arrow>
                  <Button data-testid="pipeline-current">
                    {currentPipeline}
                  </Button>
                </Dropdown>
              </Menu.Item>
              <Menu.Item key="2">
                <Search
                  nodes={graph?.nodes}
                  focusedNode={focusedNode}
                  setFocusedNode={setFocusedNode}
                  setDetailNode={setDetailNode}
                />
              </Menu.Item>
              <Menu.Item
                key="3"
                style={{ float: "right", marginLeft: "auto" }}
                onClick={() => {
                  update({})
                    .then(() => router.reload())
                    .catch(() => message.error("Failed to update!"));
                }}
              >
                <Button type="dashed" ghost={true}>
                  Update Graphs
                </Button>
              </Menu.Item>
              <Menu.Item key="4" style={{ float: "right" }}>
                <Settings animate={animate} setAnimate={setAnimate} />
              </Menu.Item>
              <Menu.Item key="5" style={{ float: "right" }}>
                Metrics refresh:&nbsp;
                <Dropdown overlay={menuRefresh}>
                  <a href={"/#"} onClick={(e) => e.preventDefault()}>
                    {refreshIntervals[refreshInterval]} <DownOutlined />
                  </a>
                </Dropdown>
              </Menu.Item>
            </Menu>
            <Spin
              style={{
                position: "fixed",
                top: "1.5em",
                right: "1em",
              }}
              indicator={<LoadingOutlined spin />}
              spinning={isLoadingMetrics}
            />
          </Header>
          <Content
            style={{
              minHeight: "100vh",
              paddingTop: "64px",
              position: "relative",
            }}
          >
            <Row style={{ position: "fixed" }}>
              {graph
                ? (
                  <GraphVisualization
                    data-testid="graph"
                    data={graph}
                    metrics={metrics}
                    refetchMetrics={() => refetchMetrics()}
                    onClickNode={(node: Node) => setDetailNode(node)}
                    width={width ? width : window.innerWidth}
                    height={height ? height - 64 : window.innerHeight - 64}
                    focusedNode={focusedNode}
                    animate={animate}
                  />
                )
                : (
                  <Alert
                    message="Error"
                    description="Could not load graph"
                    type="error"
                    showIcon
                  />
                )}
            </Row>
            <Row
              style={{
                padding: "0 50px",
                width: width,
                zIndex: 1,
                top: height ? height - 147 : 400,
                position: "absolute",
              }}
            >
              <DetailsCard node={detailNode} />
            </Row>
          </Content>
        </Layout>
      </div>
    );
  }

  return (
    <div ref={ref} className="application">
      <div className="spinningContainer">
        <Spin
          data-testid="loading"
          tip={isUpdating ? "Updating Pipelines..." : "Loading..."}
        />
      </div>
    </div>
  );
};

export default App;
