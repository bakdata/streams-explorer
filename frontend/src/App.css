.application {
  text-align: center;
  background: rgba(0, 0, 0, 0.05);
  border-radius: 4px;
}

.spinningContainer {
  padding-top: 50vh;
  min-height: 100vh;
}

.header {
  position: fixed;
  z-index: 1;
  width: 100%;
}

.ant-menu-dark.ant-menu-horizontal > .ant-menu-item:hover {
  background-color: transparent !important;
}
