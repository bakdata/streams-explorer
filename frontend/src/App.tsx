import "./App.css";
import {
  usePipelinesApiPipelinesGet,
  useGraphPositionedApiGraphGet,
  useMetricsApiMetricsGet,
} from "./api/fetchers";
import DetailsCard from "./components/DetailsCard";
import GraphVisualization from "./components/GraphVisualization";
import { graphConfig } from "./graphConfiguration";
import { DownOutlined } from "@ant-design/icons";
import {
  Layout,
  Menu,
  Dropdown,
  Button,
  Row,
  Spin,
  message,
  Alert,
} from "antd";
import { AutoComplete } from "antd";
import React, { useState, useRef, useEffect, useCallback } from "react";
import { useResizeDetector } from "react-resize-detector";
import { useMutate } from "restful-react";
import { useHistory, useLocation } from "react-router-dom";
import qs, { ParsedQs } from "qs";

const { Option } = AutoComplete;

const { Header, Content } = Layout;

const App: React.FC = () => {
  const ALL_PIPELINES = "all pipelines";
  const [currentPipeline, setCurrentPipeline] = useState(ALL_PIPELINES);
  const [detailNode, setDetailNode] = useState<string | null>(null);
  const [focusedNode, setFocusedNode] = useState<string | null>(null);
  const [searchWidth, setSearchWidth] = useState<number>(300);
  const history = useHistory();

  const location = useLocation();
  const ref = useRef<HTMLDivElement>(null!);
  const onResize = useCallback(() => {}, []);
  const { width, height } = useResizeDetector({
    targetRef: ref,
    refreshMode: "debounce",
    refreshRate: 100,
    onResize,
  });
  const defaultRefreshInterval = 30;
  const refreshIntervals: Record<number, string> = {
    0: "off",
    60: "60s",
    30: "30s",
    10: "10s",
  };
  const [refreshInterval, setRefreshInterval] = useState(
    defaultRefreshInterval
  );

  const { mutate: update, loading: isUpdating } = useMutate({
    verb: "POST",
    path: "/api/update",
  });

  const {
    data: graph,
    loading: isLoadingGraph,
    error: graphError,
  } = useGraphPositionedApiGraphGet({
    queryParams:
      currentPipeline !== ALL_PIPELINES
        ? { pipeline_name: currentPipeline }
        : undefined,
  });

  const {
    data: pipelines,
    loading: isLoadingPipelines,
    error: pipelineError,
  } = usePipelinesApiPipelinesGet({});

  const { data: metrics, refetch: refetchMetrics } = useMetricsApiMetricsGet(
    {}
  );

  useEffect(() => {
    if (refreshInterval && refreshInterval > 0) {
      const interval = setInterval(refetchMetrics, refreshInterval * 1000);
      return () => clearInterval(interval);
    }
  }, [refetchMetrics, refreshInterval]);

  // find longest node name and multiply string length by char width 8
  // doesn't cause long delays as builtin function
  useEffect(() => {
    if (graph) {
      setSearchWidth(
        Math.max(...graph.nodes.map((node) => node.label.length)) * 8
      );
    }
  }, [graph]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    const parsedQs: ParsedQs = qs.parse(location.search, {
      ignoreQueryPrefix: true,
    });
    const pipeline: string = parsedQs.pipeline as string;
    if (pipeline) {
      setCurrentPipeline(pipeline);
    }
    const focusNode: string = parsedQs.focusNode as string;
    if (focusNode) {
      setFocusedNode(focusNode);
      setDetailNode(focusNode);
    }
  }, [location]);

  if (graphError) {
    message.error(graphError.message);
    return <Spin spinning={false} tip="Failed to load graph" />;
  }
  if (pipelineError) {
    message.error(pipelineError?.message);
  }

  const menuPipeline = (
    <Menu
      onClick={(e) => {
        history.push(`/?pipeline=${e.key.toString()}`);
        setCurrentPipeline(e.key.toString());
        setFocusedNode(null);
      }}
    >
      <Menu.Item key={ALL_PIPELINES}>
        <i>{ALL_PIPELINES}</i>
      </Menu.Item>
      {pipelines?.pipelines.map((name: string) => (
        <Menu.Item key={name}>{name}</Menu.Item>
      ))}
    </Menu>
  );

  const menuRefresh = (
    <Menu
      onClick={(e) => {
        setRefreshInterval(Number(e.key));
      }}
    >
      {Object.keys(refreshIntervals).map((key: any) => (
        <Menu.Item key={key}>{refreshIntervals[key]}</Menu.Item>
      ))}
    </Menu>
  );

  if (!isLoadingGraph && !isLoadingPipelines && !isUpdating) {
    return (
      <div ref={ref} className="application">
        <Layout className="layout">
          <Header className="header" style={{ zIndex: 2 }}>
            <Menu theme="dark" mode="horizontal" selectable={false}>
              <Menu.Item key="1">
                Pipeline:&nbsp;
                <Dropdown overlay={menuPipeline} placement="bottomLeft" arrow>
                  <Button>{currentPipeline}</Button>
                </Dropdown>
              </Menu.Item>
              <Menu.Item>
                <AutoComplete
                  style={{
                    width: searchWidth,
                    maxWidth: 480,
                  }}
                  placeholder="Search Node"
                  allowClear={true}
                  defaultActiveFirstOption={true}
                  listHeight={512}
                  dropdownStyle={{
                    minWidth: searchWidth,
                  }}
                  filterOption={(inputValue, option) =>
                    option?.value
                      .toUpperCase()
                      .indexOf(inputValue.toUpperCase()) !== -1
                  }
                  onSelect={(nodeId: string) => {
                    setFocusedNode(nodeId);
                    setDetailNode(nodeId);
                    history.push(`/?focusNode=${nodeId}`);
                  }}
                >
                  {graph?.nodes.map((node) => (
                    <Option value={node.id} key={node.id}>
                      {node.id}
                    </Option>
                  ))}
                </AutoComplete>
              </Menu.Item>
              <Menu.Item style={{ float: "right" }}>
                Metrics refresh:&nbsp;
                <Dropdown overlay={menuRefresh}>
                  <a href={"/#"}>
                    {refreshIntervals[refreshInterval]} <DownOutlined />
                  </a>
                </Dropdown>
              </Menu.Item>
              <Menu.Item
                style={{ float: "right" }}
                onClick={() => {
                  update({})
                    .then(() => window.location.reload())
                    .catch(() => message.error("Failed to update!"));
                }}
              >
                <Button type="dashed" ghost={true}>
                  Update Graphs
                </Button>
              </Menu.Item>
            </Menu>
          </Header>
          <Content
            style={{
              minHeight: "100vh",
              paddingTop: "64px",
              position: "relative",
            }}
          >
            <Row style={{ position: "fixed" }}>
              {graph ? (
                <GraphVisualization
                  id="topology-graph"
                  data={graph}
                  config={graphConfig}
                  metrics={metrics}
                  refetchMetrics={() => refetchMetrics()}
                  onClickNode={(nodeId: string) => setDetailNode(nodeId)}
                  width={width}
                  height={height! - 64}
                  focusedNode={focusedNode}
                />
              ) : (
                <Alert
                  message="Error"
                  description="Could not load graph"
                  type="error"
                  showIcon
                />
              )}
            </Row>
            <Row
              style={{
                padding: "0 50px",
                width: width,
                zIndex: 1,
                top: height! - 147,
                position: "absolute",
              }}
            >
              <DetailsCard nodeID={detailNode} />
            </Row>
          </Content>
        </Layout>
      </div>
    );
  }

  return (
    <div ref={ref} className="application">
      <div className="spinningContainer">
        <Spin tip={isUpdating ? "Updating Pipelines..." : "Loading..."} />
      </div>
    </div>
  );
};

export default App;
