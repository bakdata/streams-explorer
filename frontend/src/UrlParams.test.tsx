import {
  fireEvent,
  render,
  wait,
  waitForElement,
  within,
} from "@testing-library/react";
import { createMemoryHistory } from "history";
import nock from "nock";
import React from "react";
import { Router } from "react-router";
import { useLocation } from "react-router-dom";
import { RestfulProvider } from "restful-react";
import App from "./App";

beforeAll(() => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  // disable resize observer
  (window as any).ResizeObserver = class MockResizeObserver {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

// -- Mock GraphVisualization
jest.mock("./components/GraphVisualization", () => {
  return function DummyGraphVisualization() {
    return <div data-testid="graph"></div>;
  };
});

const LocationDisplay = () => {
  const location = useLocation();
  return (
    <div>
      <div data-testid="location-pathname">{location.pathname}</div>
      <div data-testid="location-search">{location.search}</div>
    </div>
  );
};

describe("url parameters", () => {
  it("should update focus-node param", async () => {
    jest.setTimeout(30000);

    // -- Mock backend endpoints
    const nockGraph = nock("http://localhost")
      .persist()
      .get("/api/graph")
      .reply(200, {
        directed: true,
        multigraph: false,
        graph: {},
        nodes: [
          {
            id: "test-app",
            label: "test-app",
            node_type: "streaming-app",
            icon: null,
            x: 0,
            y: 0,
          },
          {
            id: "test-topic",
            label: "test-topic",
            node_type: "topic",
            icon: null,
            x: 10,
            y: 0,
          },
        ],
        edges: [
          {
            source: "test-app",
            target: "test-topic",
          },
        ],
      });

    const nockPipelineGraph = nock("http://localhost")
      .persist()
      .get("/api/graph?pipeline_name=test-pipeline")
      .reply(200, {
        directed: true,
        multigraph: false,
        graph: {},
        nodes: [
          {
            id: "test-app",
            label: "test-app",
            node_type: "streaming-app",
            icon: null,
            x: 0,
            y: 0,
          },
          {
            id: "test-topic",
            label: "test-topic",
            node_type: "topic",
            icon: null,
            x: 10,
            y: 0,
          },
        ],
        edges: [
          {
            source: "test-app",
            target: "test-topic",
          },
        ],
      });

    nock("http://localhost")
      .persist()
      .get("/api/pipelines")
      .reply(200, {
        pipelines: ["test-pipeline"],
      });

    nock("http://localhost")
      .persist()
      .get("/api/metrics")
      .reply(200, [
        {
          node_id: "test-app",
          messages_in: null,
          messages_out: null,
          consumer_lag: null,
          consumer_read_rate: null,
          topic_size: null,
          replicas: null,
          connector_tasks: null,
        },
        {
          node_id: "test-topic",
          messages_in: null,
          messages_out: null,
          consumer_lag: null,
          consumer_read_rate: null,
          topic_size: null,
          replicas: null,
          connector_tasks: null,
        },
      ]);

    const nockNode = nock("http://localhost")
      .persist()
      .get("/api/node/test-app")
      .reply(200, {
        node_id: "test-app",
        node_type: "streaming-app",
        info: [],
      });

    const history = createMemoryHistory();
    history.push({ pathname: "/", search: "?pipeline=test-pipeline" });

    const { getByTestId, getAllByTestId, getByText, asFragment } = render(
      <RestfulProvider base="http://localhost">
        <Router history={history}>
          <LocationDisplay />
          <App />
        </Router>
      </RestfulProvider>
    );
    expect(getByTestId("location-pathname")).toHaveTextContent("/");
    expect(getByTestId("location-search")).toHaveTextContent(
      "?pipeline=test-pipeline"
    );

    await waitForElement(() => getByTestId("graph"));
    expect(asFragment()).toMatchSnapshot();

    const currentPipeline = getByTestId("pipeline-current");
    expect(
      within(currentPipeline).getByText("test-pipeline")
    ).toBeInTheDocument();
    expect(nockPipelineGraph.isDone()).toBeTruthy(); // specific graph endpoint was called
    expect(nockGraph.isDone()).toBeFalsy();

    const nodeSelect = getByTestId("node-select");
    const input = within(nodeSelect).getByRole("combobox") as HTMLInputElement;
    expect(input).toHaveValue("");

    // -- set focus-node through UI
    expect(nockNode.isDone()).toBeFalsy();
    await wait(() => {
      fireEvent.change(input, { target: { value: "test-app" } });
      expect(input).toHaveValue("test-app");
      let options = getAllByTestId("node-option");
      expect(options).toHaveLength(1);
      fireEvent.click(options[0]);

      // -- check result: pipeline should be present
      expect(getByTestId("location-search")).toHaveTextContent(
        "?pipeline=test-pipeline&focus-node=test-app"
      );
      expect(nockNode.isDone()).toBeTruthy();
    });

    // -- set pipeline through UI
    fireEvent.mouseOver(getByTestId("pipeline-current"));
    await wait(() => {
      expect(getByTestId("pipeline-select")).toBeInTheDocument();
      const pipeline = getByText("all pipelines");
      expect(pipeline).toBeInTheDocument();
      fireEvent.click(pipeline);
      expect(getByTestId("location-search")).toHaveTextContent(
        "?pipeline=all pipelines"
      );
    });
  });
});

export {};
