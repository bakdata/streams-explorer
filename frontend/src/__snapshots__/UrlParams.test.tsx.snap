// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url parameters should update focus-node param 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="location-pathname"
    >
      /
    </div>
    <div
      data-testid="location-search"
    >
      ?pipeline=test-pipeline
    </div>
  </div>
  <div
    class="application"
  >
    <div
      class="spinningContainer"
    >
      <div
        class="ant-spin ant-spin-spinning ant-spin-show-text"
        data-testid="loading"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
        <div
          class="ant-spin-text"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`url parameters should update focus-node param 2`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="location-pathname"
    >
      /
    </div>
    <div
      data-testid="location-search"
    >
      ?pipeline=test-pipeline
    </div>
  </div>
  <div
    class="application"
  >
    <section
      class="ant-layout layout"
    >
      <header
        class="ant-layout-header header"
        style="z-index: 2;"
      >
        <ul
          class="ant-menu ant-menu-dark ant-menu-root ant-menu-horizontal"
          role="menu"
        >
          <li
            class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
            role="menuitem"
            style="display: none;"
          >
            <div
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              role="button"
            >
              <span>
                ···
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
          <li
            class="ant-menu-item"
            role="menuitem"
          >
            Pipeline: 
            <button
              class="ant-btn ant-dropdown-trigger"
              data-testid="pipeline-current"
              type="button"
            >
              <span>
                test-pipeline
              </span>
            </button>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
            role="menuitem"
            style="display: none;"
          >
            <div
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              role="button"
            >
              <span>
                ···
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
          <li
            class="ant-menu-item ant-menu-item-only-child"
            role="menuitem"
          >
            <div
              class="ant-select ant-select-auto-complete ant-select-single ant-select-allow-clear ant-select-show-search"
              data-testid="node-select"
              style="width: 80px; max-width: 480px;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    role="combobox"
                    type="search"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-placeholder"
                >
                  Search Node
                </span>
              </div>
            </div>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
            role="menuitem"
            style="display: none;"
          >
            <div
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              role="button"
            >
              <span>
                ···
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
          <li
            class="ant-menu-item"
            role="menuitem"
            style="float: right;"
          >
            Metrics refresh: 
            <a
              class="ant-dropdown-trigger"
              href="/#"
            >
              30s 
              <span
                aria-label="down"
                class="anticon anticon-down"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </span>
            </a>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
            role="menuitem"
            style="display: none;"
          >
            <div
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              role="button"
            >
              <span>
                ···
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
          <li
            class="ant-menu-item ant-menu-item-only-child"
            role="menuitem"
            style="float: right;"
          >
            <button
              class="ant-btn ant-btn-dashed ant-btn-background-ghost"
              type="button"
            >
              <span>
                Update Graphs
              </span>
            </button>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
            role="menuitem"
            style="visibility: hidden; position: absolute; display: none;"
          >
            <div
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              role="button"
            >
              <span>
                ···
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
        </ul>
      </header>
      <main
        class="ant-layout-content"
        style="min-height: 100vh; padding-top: 64px; position: relative;"
      >
        <div
          class="ant-row"
          style="position: fixed;"
        >
          <div
            data-testid="graph"
          />
        </div>
        <div
          class="ant-row"
          style="padding: 0px 50px; z-index: 1; top: 400px; position: absolute;"
        >
          <div
            class="details"
          >
            <div
              class="ant-card ant-card-bordered"
            >
              <div
                class="ant-card-head"
                style="background-color: rgb(74, 78, 77); color: white;"
              >
                <div
                  class="ant-card-head-wrapper"
                >
                  <div
                    class="ant-card-head-title"
                  >
                    Click on Node  to see Details
                  </div>
                </div>
              </div>
              <div
                class="ant-card-body"
              />
            </div>
          </div>
        </div>
      </main>
    </section>
  </div>
</DocumentFragment>
`;
